pkg/trait/pdb.go

```diff
@@ -0,0 +1,111 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trait
+
+import (
+	"fmt"
+	v1 "github.com/apache/camel-k/pkg/apis/camel/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// The Pbd trait allows to configure the PodDisruptionBudget resource.
+//
+// +camel-k:trait=pbd
+type pdbTrait struct {
+	BaseTrait      `property:",squash"`
+	MaxUnavailable string `property:"max-unavailable" json:"maxUnavailable,omitempty"`
+	MinAvailable   string `property:"min-available" json:"minAvailable,omitempty"`
+}
+
+func newPdbTrait() Trait {
+	return &pdbTrait{
+		BaseTrait: NewBaseTrait("pdb", 900),
+	}
+}
+
+func (t *pdbTrait) Configure(e *Environment) (bool, error) {
+	if t.Enabled != nil && *t.Enabled {
+		strategy, err := e.DetermineControllerStrategy()
+		if err != nil {
+			return false, fmt.Errorf("unable to determine the controller stratedy")
+		}
+
+		if strategy == ControllerStrategyCronJob {
+			return false, fmt.Errorf("poddisruptionbudget isn't supported with cron-job controller strategy")
+		}
+
+		if t.MaxUnavailable != "" && t.MinAvailable != "" {
+			return false, fmt.Errorf("both minAvailable and maxUnavailable can't be set simultaneously")
+		}
+
+		return e.IntegrationInPhase(
+			v1.IntegrationPhaseDeploying,
+			v1.IntegrationPhaseRunning,
+		), nil
+	}
+	return false, nil
+}
+
+func (t *pdbTrait) Apply(e *Environment) error {
+	pdb, err := t.generatePodDisruptionBudget(e)
+	if err != nil {
+		return err
+	}
+	if pdb != nil {
```

Comment:
I think we can remove that check as the `pdb` is never `nil`.

Reply:
done